consensus/core/src/transactions_client.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::block::Transaction;
+use crate::context::Context;
+use mysten_metrics::metered_channel;
+use mysten_metrics::metered_channel::channel_with_total;
+use std::sync::Arc;
+use tap::tap::TapFallible;
+use thiserror::Error;
+use tracing::error;
+
+/// Maximum number of transactions to be fetched per request of `next`
+const MAX_FETCHED_TRANSACTIONS: usize = 100;
+/// The maximum number of transactions pending to the queue to be pulled for block proposal
+const MAX_PENDING_TRANSACTIONS: usize = 2_000;
+
+/// The TransactionsConsumer is responsible for fetching the next transactions to be included for the block proposals.
+/// The transactions are submitted to a channel which is shared between the TransactionsConsumer and the TransactionsClient
+/// and are pulled every time the `next` method is called.
+#[allow(dead_code)]
+pub(crate) struct TransactionsConsumer {
+    tx_receiver: metered_channel::Receiver<Transaction>,
+    max_fetched_per_request: usize,
+}
+
+#[allow(dead_code)]
+impl TransactionsConsumer {
+    pub(crate) fn new(tx_receiver: metered_channel::Receiver<Transaction>) -> Self {
+        Self {
+            tx_receiver,
+            max_fetched_per_request: MAX_FETCHED_TRANSACTIONS,
+        }
+    }
+
+    pub(crate) fn with_max_fetched_per_request(mut self, max_fetched_per_request: usize) -> Self {
+        self.max_fetched_per_request = max_fetched_per_request;
+        self
+    }
+
+    // Attempts to fetch the next transactions that have been submitted for sequence.
+    pub(crate) fn next(&mut self) -> Vec<Transaction> {
+        let mut transactions = Vec::new();
+        while let Ok(transaction) = self.tx_receiver.try_recv() {
+            transactions.push(transaction);
+
+            if transactions.len() >= self.max_fetched_per_request {
+                break;
+            }
+        }
+        transactions
+    }
+}
+
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct TransactionsClient {
+    sender: metered_channel::Sender<Transaction>,
+}
+
+#[derive(Error, Debug)]
+pub enum ClientError {
+    #[error("Failed to submit transaction to consensus: {0}")]
+    SubmitError(String),
+}
+
+#[allow(dead_code)]
+impl TransactionsClient {
+    pub(crate) fn new(context: Arc<Context>) -> (Self, metered_channel::Receiver<Transaction>) {
+        let (sender, receiver) = channel_with_total(
+            MAX_PENDING_TRANSACTIONS,
+            &context.metrics.channel_metrics.tx_transactions_submit,
+            &context.metrics.channel_metrics.tx_transactions_submit_total,
+        );
+
+        (Self { sender }, receiver)
+    }
+
+    // Submits a transaction to be sequenced.
+    pub async fn submit(&self, transaction: Vec<u8>) -> Result<(), ClientError> {
```

Comment:
Not sure if this should accept bytes::Bytes instead. But we can iterate on this later.